paddle/fluid/framework/prune.cc

```diff
@@ -269,124 +299,154 @@ void CloneWholeBlock(proto::ProgramDesc* input, proto::ProgramDesc* output,
-      }
     }
   }
 
-  // Step 6. Copy the forward vars to new ProgramDesc
+  // Step 3. Copy the forward vars to new ProgramDesc,
   // construct all var's map before clear
-  auto* var_field = output_block->mutable_vars();
+  auto* origin_vars = origin->mutable_vars();
+  auto* pruned_vars = pruned->mutable_vars();
   std::unordered_map<std::string, proto::VarDesc> var_map;
-  for (const auto& var : *var_field) {
+  for (const auto& var : *origin_vars) {
     var_map[var.name()] = var;
   }
+  pruned_vars->Clear();
+
   std::unordered_set<std::string> var_names;
   var_names.insert(op_input_vars.begin(), op_input_vars.end());
   var_names.insert(op_output_vars.begin(), op_output_vars.end());
-  var_field->Clear();
   for (const auto& name : var_names) {
-    *var_field->Add() = var_map[name];
+    if (var_map.count(name)) {
+      *pruned_vars->Add() = var_map[name];
+    }
   }
-}
+}  // namespace framework
 
-std::unique_ptr<framework::ProgramDesc> PruneBackward(
+std::tuple<framework::ProgramDesc, std::map<int, int>> PruneBackward(
     const framework::ProgramDesc& origin) {
   // Copy original ProgramDesc, origin can't be change
   framework::ProgramDesc origin_clone(origin);
 
-  // Step 1. Update loss op's role & set loss op to be target
-  //   The loss op's op_role is (kForward | kLoss)
-  //   The input ProgramDesc should have loss operator.
-  auto ops = origin_clone.Block(0).AllOps();
-  bool has_loss_op = false;
-  for (auto op : ops) {
-    int op_role =
-        boost::get<int>(op->GetAttr(OpProtoAndCheckerMaker::OpRoleAttrName()));
-    if (op_role == (static_cast<int>(OpRole::kForward) |
-                    static_cast<int>(OpRole::kLoss))) {
-      op->SetAttr(OpProtoAndCheckerMaker::OpRoleAttrName(),
-                  static_cast<int>(OpRole::kForward));
-      op->SetIsTarget(true);
-      has_loss_op = true;
-    } else if (op_role == (static_cast<int>(OpRole::kBackward) |
-                           static_cast<int>(OpRole::kLoss))) {
-      op->SetIsTarget(false);
-      break;
+  // Step 1: check if the program contains grad loss operator.
+  // If not, the program need no pruning.
+  bool has_loss_grad_op = false;
+  std::queue<int> block_contains_loss;
+  std::queue<int> block_contains_loss_grad;
+  for (size_t i = 0; i < origin_clone.Size(); i++) {
+    auto block_ops = origin_clone.Block(i).AllOps();
+    for (auto op : block_ops) {
+      int op_role = boost::get<int>(
+          op->GetAttr(OpProtoAndCheckerMaker::OpRoleAttrName()));
+      if (op_role == (static_cast<int>(OpRole::kBackward) |
+                      static_cast<int>(OpRole::kLoss))) {
+        op->SetIsTarget(false);
+        has_loss_grad_op = true;
+        break;
+      }
     }
   }
-  PADDLE_ENFORCE_EQ(has_loss_op, true,
-                    "The Program need to be pruned its backward part"
-                    "should have loss operator.");
+  PADDLE_ENFORCE_EQ(
```

Comment:
How about make that this function can be invoked with a forward network. In `framework.py`, the original implementation use `self._appending_grad_times >= 1` to distinguish whether it is a forward network. But it is a bit hard coding...

Reply:
Done，thanks.